src/parser/jobs/blm/Leylines.js

```diff
@@ -0,0 +1,100 @@
+import React from 'react'
+import ACTIONS from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import {Rule, Requirement} from 'parser/core/modules/Checklist'
+import Module from 'parser/core/Module'
+import {ActionLink} from 'components/ui/DbLink'
+import {i18nMark, Trans} from '@lingui/react'
+
+const LEYLINE_DURATION = 30000
+
+export default class Leylines extends Module {
+	static handle = 'leylines'
+	static i18n_id = i18nMark('blm.leylines.title')
+	static title = 'Ley Lines'
+
+	static dependencies = [
+		'checklist',
+	]
+	_CircleOfPowers = {
+		current: null,
+		history: [],
+	}
+	_LeyLineHistory = []
+
+	constructor(...args) {
+		super(...args)
+		this.addHook('cast', {
+			by: 'player',
+			abilityId: ACTIONS.LEY_LINES.id,
+		}, this._onCastLeyLines)
+		this.addHook('applybuff', {
+			by: 'player',
+			abilityId: STATUSES.CIRCLE_OF_POWER.id,
+		}, this._onApplyCircleOfPower)
+		this.addHook('removebuff', {
+			by: 'player',
+			abilityId: STATUSES.CIRCLE_OF_POWER.id,
+		}, this._onRemoveCircleOfPower)
+		this.addHook('complete', this._onComplete)
+	}
+
+	_onCastLeyLines(event) {
+		this.ll = this._LeyLineHistory
+		this.ll.push(event.timestamp)
+
+	}
+
+	_onApplyCircleOfPower(event) {
+		this.cop = this._CircleOfPowers
+		if (this.cop.current) { this.cop.history.push(this.cop.current) }
+
+		this.cop.current = {
+			start: event.timestamp,
+			stop: null,
+		}
+		this._CircleOfPowers = this.cop
+	}
+
+	_onRemoveCircleOfPower(event) {
+		this.cop = this._CircleOfPowers
+		if (this.cop.current) {
+			this.cop.current.stop = event.timestamp
+		}
+		this._CircleOfPowers = this.cop
+	}
+
+	_percentFunction(numberOfLeyLines, SumOfCoPUpime) {
+		return (SumOfCoPUpime/(numberOfLeyLines*LEYLINE_DURATION))*100
+	}
+
+	_onComplete(event) {
+		this.cop = this._CircleOfPowers
+		if (!this.cop.history.length) { return }
```

Comment:
This check should probably be after the one below it - if you hit some inane edge case where the only time they were in Ley Lines was at the very end of the fight, you'd have a value in `current` to add into history, but you'd short circuit before you got the chance because there were no _other_ history events.

Alternatively, it might be worth a warning in and of itself - no Circle of Power history implies that the player didn't use Ley Lines once in the entire encounter, which is a pretty big mistake to be making.

Reply:
You are right with the first part. We do have more of the issue that CoP simmply isn't tracked in fflogs sometimes